salesforce/events.go

```diff
@@ -0,0 +1,341 @@
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.post(ctx, location, member)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := common.UnmarshalJSON[SFAPIResponseBody](resp)
+	if err != nil {
+		return nil, err
+	}
+
+	if res.Warnings != nil && len(res.Warnings) > 0 {
+		slog.Warn("CreateEventChannelMember", "warnings", res.Warnings)
+	}
+
+	member.Id = res.Id
+
+	return member, nil
+}
+
+// nolint: lll
+// https://developer.salesforce.com/docs/atlas.en-us.api_tooling.meta/api_tooling/tooling_api_objects_eventrelayconfig.htm?q=EventRelayConfig
+func (c *Connector) CreateEventRelayConfig(
+	ctx context.Context,
+	cfg *EventRelayConfig,
+) (*EventRelayConfig, error) {
+	location, err := joinURLPath(c.BaseURL, "tooling/sobjects/EventRelayConfig")
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.post(ctx, location, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := common.UnmarshalJSON[SFAPIResponseBody](resp)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg.Id = res.Id
+
+	return cfg, nil
+}
+
+// nolint: lll
+// https://developer.salesforce.com/docs/atlas.en-us.api_tooling.meta/api_tooling/tooling_api_objects_eventrelayconfig.htm?q=EventRelayConfig
+func (c *Connector) RunEventRelay(ctx context.Context, configId string) (*EventRelayConfig, error) {
+	location, err := joinURLPath(c.BaseURL, "tooling/sobjects/EventRelayConfig", configId)
+	if err != nil {
+		return nil, err
+	}
+
+	config := &EventRelayConfig{
+		Id: configId,
+		Metadata: &EventRelayConfigMetadata{
+			State: "RUN",
+		},
+	}
+
+	resp, err := c.patch(ctx, location, config)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := common.UnmarshalJSON[SFAPIResponseBody](resp)
+	if err != nil {
+		return nil, err
+	}
+
+	config.Id = res.Id
+
+	return config, nil
+}
+
+func (c *Connector) getOrganization(ctx context.Context) (map[string]*ajson.Node, error) {
+	location, err := joinURLPath(c.BaseURL, "connect/organization")
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.get(ctx, location)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.Body.GetObject()
+}
+
+func (c *Connector) GetOrganizationId(ctx context.Context) (string, error) {
+	org, err := c.getOrganization(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	return org["orgId"].MustString(), nil
+}
+
+func (c *Connector) GetRemoteResouece(ctx context.Context, channelId string) (string, error) {
+	orgId, err := c.GetOrganizationId(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("aws.partner/salesforce.com/%s/%s", orgId, channelId), nil
+}
+
+// nolint: lll
+// https://developer.salesforce.com/docs/atlas.en-us.chatterapi.meta/chatterapi/connect_resources_named_credentials_external_credentials.htm
+func (c *Connector) CreateExternalCredential(
+	ctx context.Context,
+	creds *ExternalCredential,
+) (*ExternalCredential, error) {
+	location, err := joinURLPath(c.BaseURL, "named-credentials/external-credentials")
```

Comment:
can we be consistent and use the tooling API endpoint for this? It seems like https://developer.salesforce.com/docs/atlas.en-us.api_tooling.meta/api_tooling/tooling_api_objects_namedcredential.htm does the same thing.

Reply:
I have tried to use the tooling API, but the API seems to be broken. Meaning, it doesn't recognize the parameters in the docs. I think these docs are auto-generated and they don't validate the API.